src/Security/Voter/Voter.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace HalloVerden\VoterBundle\Security\Voter;
+
+use Symfony\Bundle\SecurityBundle\Security;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AccessDecisionManagerInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\Voter as SymfonyVoter;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\Authenticator\Token\PostAuthenticationToken;
+
+abstract class Voter extends SymfonyVoter {
+
+  /**
+   * @var string[]
+   */
+  protected array $adminRoles = [];
+
+  /**
+   * AbstractEndpointVoter constructor.
+   */
+  public function __construct(
+    protected readonly Security                       $security,
+    protected readonly AccessDecisionManagerInterface $accessDecisionManager,
+  ) {
+  }
+
+  public final function vote(TokenInterface $token, mixed $subject, array $attributes): int {
+    return parent::vote($token, $subject, $attributes);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function supportsAttribute(string $attribute): bool {
+    return $attribute === static::class;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public final function supportsType(string $subjectType): bool {
+    return $subjectType === 'array' || $subjectType === 'null';
+  }
+
+  /**
+   * @param string[] $adminRoles
+   *
+   * @return void
+   */
+  public final function setAdminRoles(array $adminRoles): void {
+    $this->adminRoles = $adminRoles;
+  }
+
+  /**
+   * @return array<string|int>
+   */
+  protected function getMandatorySubjectKeys(string $attribute): array {
+    return [];
+  }
+
+  /**
+   * @param string|int $key
+   * @param mixed      $subject
+   * @param string     $attribute
+   *
+   * @return bool
+   */
+  protected function supportsSubject(string|int $key, mixed $subject, string $attribute): bool {
+    return false;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  protected final function supports(string $attribute, mixed $subject): bool {
+    return $this->supportsAttribute($attribute)
+      && (
+        (null === $subject && empty($this->getMandatorySubjectKeys($attribute)))
+        || (\is_array($subject) && $this->supportsSubjects($attribute, $subject))
+      );
+  }
+
+  /**
+   * @inheritDoc
+   */
+  protected final function voteOnAttribute(string $attribute, mixed $subject, TokenInterface $token): bool {
+    return $this->doVote($attribute, $subject ?? [], $token);
+  }
+
+  /**
+   * @param string         $attribute
+   * @param array          $subjects
+   * @param TokenInterface $token
+   *
+   * @return bool
+   */
+  protected abstract function doVote(string $attribute, array $subjects, TokenInterface $token): bool;
```

Comment:
not sure if `doVote` is a good method name? I'm open to other suggestions. 
It's basically the same as the `voteOnAttribute` method, but where subject is always an array.

Reply:
`doVote` seems perfectly fine to me :+1: